vllm/model_executor/model_loader/loader.py

```diff
@@ -347,6 +348,136 @@ def load_model(self, *, model_config: ModelConfig,
+    def _filter_subtensors(
+        tensors: Dict[str, torch.Tensor],
+    ) -> Dict[str, torch.Tensor]:
+        """
+        Filter out all tensors that share the same memory or a subset of the
+        memory of another tensor.
+        """
+        same_storage_groups = collections.defaultdict(list)
+        for key, tensor in tensors.items():
+            if tensor.numel():
+                ptr = tensor.untyped_storage().data_ptr()
+                same_storage_groups[tensor.device, ptr].append((key, tensor))
+
+        def get_end_ptr(tensor: torch.Tensor) -> int:
+            return tensor.view(-1)[-1].data_ptr() + tensor.element_size()
+
+        result = {}
+        for group in same_storage_groups.values():
+            for k, t in group:
+                a, b = t.data_ptr(), get_end_ptr(t)
+                for k2, t2 in group:
+                    if not t2.is_contiguous():
+                        continue
+                    a2, b2 = t2.data_ptr(), get_end_ptr(t2)
+                    if a < a2 or b2 < b:
+                        continue
+                    if a2 < a or b < b2 or not t.is_contiguous():
+                        break  # t2 covers strictly more memory than t.
+                    if k2 < k:
+                        # Same tensors, keep the one with the smaller key.
+                        break
+                else:
+                    result[k] = t
+        return result
+
+    def load_model(self, *, model_config: ModelConfig,
+                   device_config: DeviceConfig,
+                   lora_config: Optional[LoRAConfig],
+                   vision_language_config: Optional[VisionLanguageConfig],
+                   parallel_config: ParallelConfig,
+                   scheduler_config: SchedulerConfig) -> nn.Module:
+        from safetensors import safe_open
+
+        from vllm.distributed import get_tensor_model_parallel_rank
+        with set_default_torch_dtype(model_config.dtype):
+            with torch.device(device_config.device):
+                model = _initialize_model(model_config, self.load_config,
+                                          lora_config, vision_language_config)
+            rank = get_tensor_model_parallel_rank()
+            pattern = os.path.join(
+                model_config.model,
+                self.pattern.format(rank=rank, part="*"),
+            )
+            filepaths = glob.glob(pattern)
+            if not filepaths:
+                # TODO: support un-sharded checkpoints too
+                raise ValueError(
+                    f"Could not find checkpoint files '{pattern}', only "
+                    f"pre-sharded checkpoints are currently supported!"
+                )
+            state_dict = self._filter_subtensors(model.state_dict())
+            for path in filepaths:
+                with safe_open(path, framework="pt") as f:
+                    for key in f.keys():  # noqa: SIM118
+                        state_dict[key].copy_(f.get_tensor(key))
```

Comment:
```
state_dict[key][:f.get_tensor(key).shape[0]].copy_(f.get_tensor(key))
```
I think you need to do this to be compatible when `enable_lora` is true.

Reply:
similar to https://github.com/vllm-project/vllm/blob/e254497b66dcd87038969b0ad34d34425edfc5fe/vllm/model_executor/layers/vocab_parallel_embedding.py#L87

Reply:
thanks for catching this! yeah there seems to be some issues with this pr in supporting lora. basically, during initialization, two things happen in vllm in this order:

1. The model loader is called to load the base model weights.
2. Lora is initialized, which may rename some parameters and reshape some weights.

currently, the worker states are saved after step (2). then, when later loading it through the model loader in step (1), it will contain some unexpected parameter names and shapes.

instead, the sharded weights should be both saved and loaded between (1) and (2), or both after step (2). the former might look something like

```python
llm = LLM(model_path, save_sharded_states=path_to_shards, ...)
```

then when loading

```python
llm = LLM(model_path, load_format="sharded_state",
          model_loader_extra_config=ShardedStateConfig(path_to_shards), ...)
```

and the latter might look like:

```python
llm = LLM(model_path, ...)

llm.save_sharded_states(path_to_shards)
```

then when loading

```python
llm = LLM(model_path, load_format="dummy", ...)

llm.load_sharded_states(path_to_shards)
```

anyone have any thoughts?

Reply:
I think the shape trick should be fine. Any post-initialization modifications can only extend existing tensors or add new ones. There should not be a situation where the shape or information stored in a tensor is different. I think we will be fine with just a warning (and comments) in loading code when that happens.

Reply:
done. btw, any small model that supports lora that can be used for test? lors is not supported for opt.

Reply:
we are using llama 2 7b with lora for testing. it's not the smallest, but that's what we got right now :(

Reply:
should be all good now. thanks for the comments!